### PR TITLE
net: tcp: Implement Nagle's algorithm

### LINKED ISSUE #30365: TCP2 does not implement Nagle algorithm
**Is your enhancement proposal related to a problem? Please describe.**
Currently TCP Nagle algorithm is not implemented when sending data. https://en.wikipedia.org/wiki/Nagle%27s_algorithm

Implementing Nagle would improve network utilization as it collects smaller TCP segments into larger ones and then send one larger segment instead of multiple smaller ones.

### PR DESCRIPTION
**Pull request by ssharks**
To improve the performance with small chunks send, implement Nagle's
algorithm. Provide the option TCP_NODELAY to disable the algorithm.

Fixes #30365.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

### CODE REVIEW

File: subsys/net/ip/tcp.c [hunk 1]
```diff
@@ -1097,6 +1128,19 @@ static int tcp_send_queued_data(struct tcp *conn)
 	}
 
 	while (tcp_unsent_len(conn) > 0) {
+		/* Implement Nagle's algorithm */
+		if ((conn->tcp_nodelay == false) && (conn->unacked_len > 0)) {
+			/* If there is already pending data */
+			if (tcp_unsent_len(conn) < conn_mss(conn)) {
+				/* The number of bytes to be transmitted is less then an MSS,
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:1135]: Typo s/then/than/

File: subsys/net/ip/tcp.c [hunk 2]
```diff
@@ -848,6 +848,37 @@ static int ip_header_add(struct tcp *conn, struct net_pkt *pkt)
...
 }
 
+static int set_tcp_nodelay(struct tcp *conn,
+				const void *value, size_t len)
+{
+	if (len != sizeof(int)) {
+		return -EINVAL;
+	}
+
+	int no_delay_int = *(int *)value;
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:858]: Minor nit, we try to introduce the variables at the beginning of the block.

**Review by jukkar — COMMENTED** [review]
Looks good, minor style issue.

File: subsys/net/ip/tcp.c [hunk 3]
```diff
@@ -848,6 +848,37 @@ static int ip_header_add(struct tcp *conn, struct net_pkt *pkt)
...
+		return -EINVAL;
+	}
+
+	conn->tcp_nodelay = (bool)no_delay_int;
+
+	return 0;
+}
+
+static int get_tcp_nodelay(struct tcp *conn,
+				void *value, size_t *len)
```
Comment (rlubos) [inline subsys/net/ip/tcp.c:870]: ditto

File: subsys/net/ip/tcp.c [hunk 4]
```diff
@@ -2994,6 +3039,56 @@ uint16_t net_tcp_get_supported_mss(const struct tcp *conn)
 	return 0;
 }
 
+int net_tcp_set_option(struct net_context *context,
+			   enum tcp_conn_option option,
+			   const void *value, size_t len)
```
Comment (rlubos) [inline subsys/net/ip/tcp.c:3044]: Please align the arguments to the opening brackets.

File: subsys/net/ip/tcp.c [hunk 5]
```diff
@@ -2994,6 +3039,56 @@ uint16_t net_tcp_get_supported_mss(const struct tcp *conn)
...
+	}
+
+	k_mutex_unlock(&conn->lock);
+
+	return ret;
+}
+
+int net_tcp_get_option(struct net_context *context,
+			    enum tcp_conn_option option,
+			    void *value, size_t *len)
```
Comment (rlubos) [inline subsys/net/ip/tcp.c:3069]: Same here.

File: subsys/net/ip/tcp_internal.h [hunk 6]
```diff
@@ -367,6 +371,57 @@ void net_tcp_init(void);
...
+/**
+ * @brief Set tcp specific options of a socket
+ *
+ * @param context Network context
+ *
+ * @return 0 on success, -EINVAL if the value is not allowed
+ */
+#if defined(CONFIG_NET_NATIVE_TCP)
+int net_tcp_set_option(struct net_context *context,
+			   enum tcp_conn_option option,
```
Comment (rlubos) [inline subsys/net/ip/tcp_internal.h:383]: And in this file as well.

File: subsys/net/lib/sockets/sockets.c [hunk 7]
```diff
@@ -1852,6 +1852,13 @@ int zsock_getsockopt_ctx(struct net_context *ctx, int level, int optname,
 			}
 			break;
 		}
+	case IPPROTO_TCP:
+		switch (optname) {
+		case TCP_NODELAY:
+			ret = net_tcp_get_option(ctx,
+				TCP_OPT_NODELAY, optval, optlen);
```
Comment (rlubos) [inline subsys/net/lib/sockets/sockets.c:1859]: Could align the arguments here as well, see how it's done above.

**Review by rlubos — COMMENTED** [review]
Just some nitpicking

File: subsys/net/ip/tcp.c [hunk 8]
```diff
@@ -848,6 +848,37 @@ static int ip_header_add(struct tcp *conn, struct net_pkt *pkt)
 	return -EINVAL;
 }
 
+static int set_tcp_nodelay(struct tcp *conn,
+				const void *value, size_t len)
```
[the remark this answers is not in the document]
Comment (ssharks) [reply subsys/net/ip/tcp.c:852]: Do you know what the max line length allowed is?

[the remark this answers is not in the document]
Comment (ssharks) [reply subsys/net/ip/tcp.c:852]: Apparently 100 characters...

File: subsys/net/ip/tcp.c (hunk 3 above)
Comment (ssharks) [reply subsys/net/ip/tcp.c]: Fixed

File: subsys/net/ip/tcp.c (hunk 4 above)
Comment (ssharks) [reply subsys/net/ip/tcp.c]: Fixed

File: subsys/net/ip/tcp.c (hunk 5 above)
Comment (ssharks) [reply subsys/net/ip/tcp.c]: Fixed

File: subsys/net/ip/tcp_internal.h (hunk 6 above)
Comment (ssharks) [reply subsys/net/ip/tcp_internal.h]: When I do so it looks like I cat a line length warning of over 100 characters, although the lines do not exceed 70 characters. Do you know what is wrong?

File: subsys/net/lib/sockets/sockets.c (hunk 7 above)
Comment (ssharks) [reply subsys/net/lib/sockets/sockets.c:1859]: Fixed

File: subsys/net/ip/tcp.c (hunk 2 above)
Comment (ssharks) [reply subsys/net/ip/tcp.c]: Fixed

File: subsys/net/ip/tcp.c (hunk 1 above)
Comment (ssharks) [reply subsys/net/ip/tcp.c]: Fixed

File: subsys/net/ip/tcp.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (rlubos) [reply subsys/net/ip/tcp.c]: 100 char/line, some folks still stick to 80 char/line as it used to be the limit until recently.

File: subsys/net/ip/tcp_internal.h (hunk 6 above)
Comment (rlubos) [reply subsys/net/ip/tcp_internal.h]: I bet our editor is configured to use 4 spaces/tab, Zephyr uses 8
image

Comment (ssharks) [reply subsys/net/ip/tcp_internal.h]: That explains, issue fixed

**Review by rlubos — APPROVED** [review]
LGTM

**Review by jukkar — APPROVED** [review]

**Pull request merged**